### ISSUE TITLE
Feature request: Support for large model sizes

### ISSUE THREAD

**Issue by mgharamti** [NCAR/DART#292]
### Use case  
I'm working with a 100th of a degree MITgcm and NBLING (an ecosystem model). The size of my state is 2,408,000,000 (i.e., ~2.4 billions). When running model_mod_check, I realized that the size of my state from TEST 1 is negative. So, I figured we need a bigger integer for the variable: ``model_size``. I added an ``i16`` (= ``SELECTED_INT_KIND(32)``) in types_mod and assigned that to ``model_size``. I passed TEST 1 and the correct ``model_size`` was written out. However, TEST 2 failed when reading my input netcdf file. Here is the ERROR MESSAGE I got:
``At line 861 of file ../../../assimilation_code/modules/io/direct_netcdf_mod.f90
Fortran runtime error: Index '-2090967295' of dimension 1 of array 'var_block' outside of expected range (1:2408000000)``

### Is your feature request related to a problem?  
Not really. I don't think we've worked with this large state size before. 

### Describe your preferred solution  
I can continue digging and adjusting variable declarations but I'm not sure this is the right strategy. Knowing the model_size and accessing the state vector is something we do everywhere inside DART. I'm looking for recommendations if there is an easy solution to bypass this. 

### Describe any alternatives you have considered  
Maybe some sort of integer aliasing in types_mod (set ``i2``, ``i4``, .. to ``i16``)? My heart tells me: "**DON'T DO THAT**"

**Comment by nancycollins** [comment]
are these the same 2 problems helen has been working on recently?  1. when the state vector size is > a 32 bit int.   we default to using i4, so i8 should be sufficient for this shouldn't it?  you shouldn't need i16 (unless my math is way off?)  
also 2. she's been looking at netcdf files and how we create them when the individual arrays are large and when the file itself needs 64 bit offsets to do random access.  (apologies if i'm mistaken here.)

**Comment by hkershaw-brown** [comment]
Hi Moha, I have a pull request in for this problem.  It has not made it to the main branch yet.
The code is on this branch: https://github.com/NCAR/DART/tree/model_size_i8

You're heart is correct, that integer aliasing in types_mod is probably a headache. 

You can take a look at the original bug report here: #278.  A user reported problems when using WRF with a model size of 3796738754 - so the same integer problems you are hitting. 

I think you should be able to grab the changes from the model_size_i8 branch and be good to go.  But, you might hit problems with large static data. Shoot me an email if you hit something unexpected or if the i8 still doesn't fix your problems.

Cheers,
Helen

**Comment by nancycollins** [comment]
maybe related to this - there is a namelist item in the i/o code:

```
&state_vector_io_nml
   buffer_state_io         = .false.
```
which does i/o and distribution to other tasks array by array instead of reading/writing the entire state into a single buffer. 
 this may help if you run into i/o problems.

**Comment by nancycollins** [comment]
fun historical fact - we changed state vector indices from i4 to i8 when 10th degree pop was one of our target models several years ago.  at that time we believed that the pop state vector size was > i4 and by testing 10th degree pop we had validated that indices and sizes of i8s were correctly supported.  

apparently no one did the math (myself included).

it turns out the pop state was just a smidge smaller than needed to exceed 32 bit indices and offsets.  the compiler complained about integer type mismatches (i4 vs i8) and we fixed those but, unbeknownst to us, we didn't actually run anything that actually tested the code.   helen's been dealing with a WRF user who finally did exceed the i4 size and now it looks like moha has, too.  

i guess the good news is that most of the system already uses i8s for state vector indices and sizes, but there were still several places where changes were needed which were not caught by testing.  helen's been working on finding and fixing those on the i8 branch.

my guess is that model_mod_check.f90 also needs some updates and the MITgcm model_mod.f90 may as well since size/index problems won't have been caught by testing a large WRF filter run.  (helen and moha should coordinate on who changes what.)
